src/logic/algo/solver.ts

```diff
@@ -23,20 +23,124 @@ export type StepToImplementName = keyof Solver;
 export class MonteCarloTreeSearch implements Solver {
   private kernel: NonNullable<MonteCarloTreeSearchHyperParams["kernel"]> = ubcKernel;
 
+  private nodeRewards: Map<TreeNode, number> = new Map();
+
+  private nodeVisits: Map<TreeNode, number> = new Map();
+
+  private nodeChildren: Map<TreeNode, Array<TreeNode>> = new Map();
+
   run(root: TreeNode, hyperParams: MonteCarloTreeSearchHyperParams): TreeNode {
+    while (true) {
+      for (let i = 0; i < hyperParams.numIterations; i += 1) {
+        this.singleRun(root, hyperParams);
+      }
+      root = this.choose(root);
+      if (root.children.length === 0) {
+        return root;
+      }
+    }
+  }
+
+  singleRun(root: TreeNode, hyperParams: MonteCarloTreeSearchHyperParams): void {
     this.kernel = hyperParams.kernel ?? this.kernel;
-    return new TreeNode();
+    const path = this.select(root);
+    const leaf = path[path.length - 1];
+    this.expand(leaf);
+    let reward = 0;
+    for (let i = 0; i < hyperParams.numRollout; i += 1) {
+      reward += this.simulate(leaf);
+    }
+    this.backup(path, reward);
   }
 
-  choose(node: TreeNode): null {
-    return null;
+  select(node: TreeNode): Array<TreeNode> {
+    const path: Array<TreeNode> = [];
+    while (true) {
+      path.push(node);
+      if (!this.nodeChildren.has(node)) {
+        return path;
+      }
+      if (this.nodeChildren.get(node) ?? [].length === 0) {
+        return path;
+      }
+      let unexploredNodes: Array<TreeNode> = this.nodeChildren.get(node) ?? [];
+      unexploredNodes = unexploredNodes.filter(
+        (unexploredNode) => !this.nodeChildren.has(unexploredNode)
+      );
+      if (unexploredNodes.length > 0) {
+        const n: TreeNode | undefined = unexploredNodes.pop();
+        if (n) {
+          path.push(n);
+        }
+        return path;
```

Comment:
Milion ifów.. Czy macie może pomysł jak to zrefaktorować?

Reply:
Chyba można bez tego 1. if'a
Czyli
```
const n: TreeNode | undefined = unexploredNodes.pop();`
```
![image](https://user-images.githubusercontent.com/21079319/116482206-fb49cf80-a884-11eb-8b86-ee30d5b867b2.png)

Reply:
Coś nie idzie :( Ale to nic, zostawię jak było , tylko podmienię "n" na coś :D